lua/bson_lua_wrapper.c

```diff
@@ -0,0 +1,372 @@
+        lua_pop(L, 1); //Stack: [bson_table, {type, value}]
+
+        lua_getfield(L, -1, "value"); //Stack: [bson_array, {type, value}, value]
+        switch ((element_type)type) {
+            case TYPE_DOCUMENT: {
+                if (!lua_istable(L, -1)) {
+                    snprintf(errorMessage, 255, "Expected table, but value was of type: %s", 
+                             lua_typename(L, lua_type(L, -1)));
+                    return -1;
+                }
+                BsonObject value;
+                bson_object_initialize_default(&value);
+
+                int success = table_to_bson_object(L, &value, errorMessage);
+                if (success != 0) {
+                    return success;
+                }
+
+                bson_array_add_object(arr, &value);
+                break;
+            }
+            case TYPE_ARRAY: {
+                if (!lua_istable(L, -1)) {
+                    snprintf(errorMessage, 255, "Expected array, but value was of type: %s", 
+                             lua_typename(L, lua_type(L, -1)));
+                    return -1;
+                }
+                BsonArray value;
+                bson_array_initialize(&value, 4);
+
+                int success = table_to_bson_array(L, &value, errorMessage);
+                if (success != 0) {
+                    return success;
+                }
+
+                bson_array_add_array(arr, &value);
+                break;
+            }
+            case TYPE_INT32: {
+                lua_Integer value = luaL_checkinteger(L, -1);
+                bson_array_add_int32(arr, (int32_t)value);
+                break;
+            }
+            case TYPE_INT64: {
+                lua_Integer value = luaL_checkinteger(L, -1);
+                bson_array_add_int64(arr, (int64_t)value);
+                break;
+            }
+            case TYPE_STRING: {
+                char *value = strdup(luaL_checkstring(L, -1));
+                bson_array_add_string(arr, value);
+                free(value);
+                break;
+            }
+            case TYPE_DOUBLE: {
+                lua_Number value = luaL_checknumber(L, -1);
+                bson_array_add_double(arr, (double)value);
+                break;
+            }
+            case TYPE_BOOLEAN: {
+                if (!lua_isboolean(L, -1)) {
+                    snprintf(errorMessage, 255, "Expected boolean, but value was of type: %s", 
+                             lua_typename(L, lua_type(L, -1)));
+                    return -1;
+                }
+                int value = lua_toboolean(L, -1);
+
+                bson_array_add_bool(arr, (bson_boolean)value);
+                break;
+            }
+            default: {
+                snprintf(errorMessage, 255, "Error parsing table, unrecognized type: %i", (int)type);
+                return -1;
+            }
+        }
+        lua_pop(L, 2); //Stack: [bson_array]
+    }
+    lua_pop(L, 1); //Stack: []
```

Comment:
@jacobkeeler Function `table_to_bson_array()` is used recursively and should leave lua stack state as it was before its call, in other way it can lead to errors during recursion (for example: some data on lua stack is needed after recursed call of function, but it was poped from stack during recursed call of function)

So I would propose remove line `lua_pop(L, 1); //Stack: []`at all.